nucypher/acumen/perception.py

```diff
@@ -30,141 +33,324 @@
+            return '{checksum} ⇀{nickname}↽ {icon}'.format(icon=self.nickname.icon,
+                                                           nickname=self.nickname,
+                                                           checksum=self.checksum[:7])
+        else:
+            return 'No Known Nodes'
+
+
+class ArchivedFleetState(BaseFleetState):
+
+    def __init__(self, checksum, nickname, timestamp, population):
+        self.checksum = checksum
+        self.nickname = nickname
+        self.timestamp = timestamp
+        self.population = population
+
+    def abridged_details(self):
+        return {"nickname": self.nickname.payload() if self.nickname is not None else NO_KNOWN_NODES,
+                "updated": self.timestamp.rfc2822()}
+
+
+# Assumptions we're based on:
+# - Every supplied node object, after its constructor has finished,
+#   has a ``.checksum_address`` and ``bytes()`` (metadata)
+# - checksum address or metadata does not change for the same Python object
+# - ``this_node`` (the owner of FleetSensor) may not have a checksum address initially
+#   (when the constructor is first called), but will have one at the time of the first
+#   `record_fleet_state()` call. This applies to its metadata as well.
+# - The metadata of ``this_node`` **can** change.
+# - For the purposes of the fleet state, nodes with different metadata are considered different,
+#   even if they have the same checksum address.
+
+class FleetState(BaseFleetState):
+
+    @classmethod
+    def new(cls, this_node=None, nodes=[]):
+        this_node_ref = weakref.ref(this_node) if this_node is not None else None
+        # Using empty checksum so that JSON library is not confused.
+        # Plus, we do need some checksum anyway. It's a legitimate state after all.
+        return cls(checksum=keccak_digest(b"").hex(),
+                   nodes={},
+                   this_node_ref=this_node_ref,
+                   this_node_metadata=None)
+
+    def __init__(self, checksum, nodes, this_node_ref, this_node_metadata):
+        self.checksum = checksum
+        self.nickname = None if checksum is None else Nickname.from_seed(checksum, length=1)
+        self._nodes = nodes
+        self.timestamp = maya.now()
+        self._this_node_ref = this_node_ref
+        self._this_node_metadata = this_node_metadata
+
+    def archived(self):
+        return ArchivedFleetState(checksum=self.checksum,
+                                  nickname=self.nickname,
+                                  timestamp=self.timestamp,
+                                  population=self.population)
+
+    def with_updated_nodes(self, new_nodes, marked_nodes):
+
+        # Checking if the node already has a checksum address
+        # (it may be created later during the constructor)
+        # or if it mutated since the last check.
+        if self._this_node_ref is not None and getattr(self._this_node_ref(), 'finished_initializing', False):
```

Comment:
While I like the way this logic is evolving, these weakref gymnastics aren't really doing it for me.  Is this all just to avoid having to lug around `additional_nodes_to_track`?

Reply:
Not really. There are several issues being solved here.

1. Circular references. The current code with `additional_nodes_to_track` has them too: `Ursula` -> `known_nodes` -> `additional_nodes_to_track` -> `Ursula`. Hence the weakrefs.

2. This code can be called at `Ursula` construction time, when its full metadata is still not available. A check for the local node to be available (currently a very awkward `getattr` of `finished_initializing`) removes the need in an additional mutating call that you have to remember to invoke sometime after you create the `Ursula`.

3. The local node can be mutated anytime. So we have to request the new metadata every time the state is updated.

Reply:
None of these are strong enough to justify the enormous hit in readability, IMO.

`weakref` is something best saved for when it's truly needed, like when it's the only way out of a hairy performance bottleneck with an object whose representation is too large (in memory) to justify doing some other way.

Reply:
> None of these are strong enough to justify the enormous hit in readability, IMO.

I wouldn't call two dereferences of a weakref an "enormous hit in readability". And the weakref is there only for dealing with the first point; if you get rid of it, the code will remain pretty much the same because there are still points 2 and 3 to worry about. Now if we prohibit node mutation, *that* will improve the readability.

> weakref is something best saved for when it's truly needed, like when it's the only way out of a hairy performance bottleneck with an object whose representation is too large (in memory) to justify doing some other way.

`Ursula` is the main object being held by the cycle here, and it's pretty large. Of course, it mainly matters for testing, but we've already hit a similar problem not so long ago. 

Personally, I believe that it is better to take a slight readability hit and use a weakref than debug a problem caused by a reference cycle half a year later.

Reply:
Related discusson: https://ptb.discord.com/channels/411401661714792449/411401661714792451/772182575237431306